crates/bevy_app/src/app.rs

```diff
@@ -823,19 +827,43 @@ impl App {
     /// # }
     /// App::new().add_plugin(bevy_log::LogPlugin::default());
     /// ```
+    ///
+    /// # Panics
+    ///
+    /// Panics if the plugin was already added to the application.
     pub fn add_plugin<T>(&mut self, plugin: T) -> &mut Self
     where
         T: Plugin,
     {
-        self.add_boxed_plugin(Box::new(plugin))
+        match self.add_boxed_plugin(Box::new(plugin)) {
+            Ok(app) => app,
+            Err(AppError::DuplicatePlugin { plugin_name }) => panic!(
+                "Error adding plugin {}: : plugin was already added in application",
+                plugin_name
+            ),
+        }
     }
 
     /// Boxed variant of `add_plugin`, can be used from a [`PluginGroup`]
-    pub(crate) fn add_boxed_plugin(&mut self, plugin: Box<dyn Plugin>) -> &mut Self {
+    pub(crate) fn add_boxed_plugin(
+        &mut self,
+        plugin: Box<dyn Plugin>,
+    ) -> Result<&mut Self, AppError> {
         debug!("added plugin: {}", plugin.name());
+        if plugin.is_unique()
+            && self
+                .plugin_registry
+                .iter()
+                .map(|plugin| plugin.name())
+                .any(|name| name == plugin.name())
```

Comment:
While I don't expect users to be adding tens of thousands of plugins, this seems like it take quadratically more time as we add more plugins to an app. Could we not separate this out into a `HashSet` of unique plugin names?

Reply:
I agree, but I'm not a fan of adding that unless it's really a problem we think we'll have

Reply:
I like the `HashSet` of plugin names because a) it'll be useful for this + plugin dependencies and b) it's great debug info for `App`. That said, I'm fine to wait on it for this PR.

Reply:
done 👍